src/bindings/c/ov/src/ov_c_api.cpp

```diff
@@ -613,47 +618,91 @@ bool ov_model_is_dynamic(const ov_model_t* model) {
     }
     return result;
 }
 
-ov_status_e ov_partial_shape_init(ov_partial_shape_t* partial_shape, const char* str) {
-    if (!partial_shape || !str) {
+ov_status_e ov_partial_shape_init(ov_partial_shape_t** partial_shape_obj, const char* str) {
+    *partial_shape_obj = nullptr;
+    if (!str) {
         return ov_status_e::GENERAL_ERROR;
     }
+    ov_partial_shape_t* partial_shape = nullptr;
     try {
         std::string s = str;
-        std::regex reg("[^,0-9.?][^\\-1]");
+        std::regex reg("[^,0-9.?}{}][^\\-1]");
         bool res = std::regex_search(s, reg);
         if (res) {
             return ov_status_e::PARAMETER_MISMATCH;
         }
-        std::vector<char*> result = split(s);
-        partial_shape->ranks = result.size();
-        if (partial_shape->ranks > MAX_DIMENSION) {
-            return ov_status_e::GENERAL_ERROR;
+
+        partial_shape = new ov_partial_shape_t;
+        if (s == "?") {  // dynamic rank
+            partial_shape->rank = ov::Dimension::dynamic();
+        } else {  // static rank
+            std::vector<char*> result = split(s);
+            size_t cnt = result.size();
+            if (cnt > MAX_DIMENSION) {
+                delete partial_shape;
+                return ov_status_e::GENERAL_ERROR;
+            }
+            partial_shape->rank = ov::Dimension(cnt);
+            for (auto& item : result) {
+                std::string dim = item;
+                if (dim == "?" || dim == "-1") {
+                    partial_shape->dims.emplace_back(ov::Dimension::dynamic());
+                } else {
+                    const std::string range_divider = "..";
+                    size_t range_index = dim.find(range_divider);
+                    if (range_index != std::string::npos) {
+                        std::string min = dim.substr(0, range_index);
+                        std::string max = dim.substr(range_index + range_divider.length());
+                        partial_shape->dims.emplace_back(min.empty() ? 0 : std::stoi(min),
+                                                         max.empty() ? ngraph::Interval::s_max : std::stoi(max));
+                    } else {
+                        partial_shape->dims.emplace_back(std::stoi(dim));
+                    }
+                }
+            }
         }
-        std::copy_n(result.begin(), result.size(), partial_shape->dims);
-    } CATCH_OV_EXCEPTIONS
+    }
+    CATCH_OV_EXCEPTIONS
+    *partial_shape_obj = partial_shape;
     return ov_status_e::OK;
 }
 
 const char* ov_partial_shape_parse(ov_partial_shape_t* partial_shape) {
     if (!partial_shape) {
         return str_to_char_array("error");
     }
-    std::string str;
-    for (int i = 0; i < partial_shape->ranks; ++i) {
-        std::string tmp = partial_shape->dims[i];
-        str += tmp;
-        if (i != partial_shape->ranks - 1)
+
+    // dynamic rank
+    if (partial_shape->rank.is_dynamic()) {
+        return str_to_char_array("?");
```

Comment:
In all statements like this one you can just return like this:
```suggestion
        return "?";
```
Otherwise you're creating a string from "?" and the return statement converts this string back to `const char*` which seems redundant.

Reply:
It will be unsafe to return a string data without allocating memory for it.